util-app/src/main/scala/com/twitter/app/App.scala

```diff
@@ -105,7 +105,16 @@ trait App extends Closable with CloseAwaitably {
   /**
    * Close `closable` when shutdown is requested. Closables are closed in parallel.
    */
-  final def closeOnExit(closable: Closable): Unit = exits.add(closable)
+  final def closeOnExit(closable: Closable): Unit = {
+    exits.add(closable)
+
+    // We check `closeDeadline` here to ensure that `closable` was added to
+    // `exits` prior to `close()` getting called. If `close()` has already been
+    // called, we should just close `closable` here.
+    if (closeDeadline != Time.Top) {
```

Comment:
uhhhh this is getting really deep in the concurrency weeds but volatile isn't really strong enough here.  the JVM could reorder this in a dirty way:

t1:
```
if (closeDeadline != Time.top)
```

t2: 
```
closeDeadline = Time.now + 5.seconds
closables.foreach(_.close(deadline))
```

t1:
```
exits.add(closable)
```
Pretty good explanation [here][0].

This probably isn't an issue in practice, but I'd rather program against the JMM than against an implementation of the JMM.  Let's just switch this to use mutexes.

[0]: https://shipilev.net/blog/2016/close-encounters-of-jmm-kind/#wishful-hb-actual